feat(PeerGrouping): Pair students with different ideas

hirokiterashima opened the pull request:
## Changes
- Add ```differentIdeas("nodeId", "componentId")``` term in PeerGrouping.logic that pairs students based on different ideas
- Refactor code

## Test Prep
1. Update database
   - ```alter table peer_groupings modify logic varchar(255) null;```
   - ```update peer_groupings set logic="manual" where logic = "[{\"name\":\"manual\"}]";```
2. Create a CRater component with detected ideas ("ComponentX"), e.g. "MusicalInstruments_Speed_score-ki_idea-ki_nonscor"
3. Add new PeerGrouping to unit content (via JSON editing), and set the logic's nodeId and componentId to point to ComponentX
```
"peerGroupings": [
...
  {
    "logic": "differentIdeas(\"NODE_ID\",\"COMPONENT_ID\")",
    "maxMembershipCount": 2,
    "name": "Different Idea",
    "tag": "dg0uytsun8"
  }
...
]
``` 
4. Create a new component that uses PeerGroup (e.g. PeerChat, Show Group Work) and choose the above "Different Idea" as its Grouping Logic

## Test DifferentIdeas Logic
- When the student gets to the new component, they should be paired into PeerGrouping with another workgroup if they have at least one different idea. The pairing should maximize different ideas if possible. 

```
Workgroup1: A, B, C
Workgroup2: A, B, C, D
Workgroup3: B, C, D, E
```
In this case, Workgroup1 should be paired with Workgroup3 since they have 2 different ideas.

## Test Random Logic
- Random logic should work as before. The related code has been refactored, but no new feature was added as far as this logic is concerned.

Closes #179

geoffreykwan commented on src/main/java/org/wise/portal/domain/peergrouping/logic/DifferentIdeasLogic.java:
```diff
@@ -0,0 +1,27 @@
+  private String nodeId;
+
+  public DifferentIdeasLogic(String logic) {
+    Pattern pattern = Pattern.compile("\"(\\w+)\"");
```
We might be able to change the regex to something like ```differentIdeas\("(\w+)",\s*"(\w+)"\)```

So the code might be something like this (I'm not sure if the escaping is correct).
```Pattern pattern = Pattern.compile("differentIdeas\\(\"(\\w+)\",\\s*\"(\\w+)\"\\)");```

geoffreykwan commented on src/main/java/org/wise/portal/service/peergroup/impl/PeerGroupServiceImpl.java:
```diff
@@ -23,142 +23,69 @@
-        workgroupsNotInPeerGroup.add(workgroup);
-      }
+    if (peerGroupThresholdService.isThresholdSatisfied(peerGrouping, workgroup.getPeriod())) {
+      return peerGrouping.getLogic().equals("random")
+        ? randomLogicService.createPeerGroup(workgroup, peerGrouping)
+        : differentIdeasLogicService.createPeerGroup(workgroup, peerGrouping);
```
Should we explicitly check for differentIdeas logic? If anything, maybe we should default to random instead of differentIdeas.

hirokiterashima replied:
Does it matter which case we test? There are only two possible logic options at this point, "random" and "differentIdeas(nodeId, componentId)" ("manual" was already handled prior to this function being called). So if it's not "random", it must be "differentIdeas". If the logic is not one of "manual", "random", or "differentIdeas", I don't think this function should be called. If we test for "differentIdeas", we have to run a Matcher (which is more expensive than an "equals" check), and "differentIdeas" occurs less often than "random", so wouldn't we want to avoid testing for it?

geoffreykwan commented on src/main/java/org/wise/portal/service/peergroup/impl/WorkgroupWithDifferentIdeas.java:
```diff
@@ -0,0 +1,27 @@
+
+  @Override
+  public int compareTo(WorkgroupWithDifferentIdeas o) {
+    if (this.numDifferentIdeas > o.numDifferentIdeas) {
+      return -1;
+    } else if (this.numDifferentIdeas < o.numDifferentIdeas) {
+      return 1;
+    } else {
+      return 0;
+    }
+  }
```
Can this be changed to ```return o.numDifferentIdeas - this.numDifferentIdeas```

geoffreykwan commented on src/main/java/org/wise/portal/service/peergrouping/logic/impl/DifferentIdeasLogicServiceImpl.java:
```diff
@@ -0,0 +1,121 @@
+
+  private List<Annotation> getIdeaDetectedAnnotations(PeerGrouping peerGrouping, Group period) {
+    DifferentIdeasLogic logic = new DifferentIdeasLogic(peerGrouping.getLogic());
+    List<Annotation> annotations = annotationDao
+        .getAnnotations(peerGrouping.getRun(), logic.getNodeId(), logic.getComponentId()).stream()
+        .filter(annotation -> annotation.getPeriod().equals(period)
+            && annotation.getType().equals("autoScore"))
+        .collect(Collectors.toList());
```
Should we call annotationDao.getAnnotationsByParams() which can filter by period id and type?

geoffreykwan commented on src/main/java/org/wise/portal/service/peergrouping/logic/impl/DifferentIdeasLogicServiceImpl.java:
```diff
@@ -0,0 +1,121 @@
+
+  private boolean hasEnoughUnpairedMembersIdeasAnnotation(PeerGrouping peerGrouping,
+      List<Annotation> annotations, Set<Workgroup> workgroupsNotInPeerGroup) {
+    return workgroupsNotInPeerGroup.stream()
+        .filter(workgroup -> hasWorkgroupAnnotation(workgroup, annotations)).count() >= 2;
```
Should the 2 be hard coded or should we look at maxMembershipCount?

geoffreykwan commented on src/main/java/org/wise/portal/service/peergrouping/logic/impl/DifferentIdeasLogicServiceImpl.java:
```diff
@@ -0,0 +1,121 @@
+
+  private Optional<Annotation> getWorkgroupAnnotation(Workgroup workgroup,
+      List<Annotation> annotations) {
+    return annotations.stream().filter(annotation -> annotation.getToWorkgroup().equals(workgroup))
+        .findFirst();
```
I think the annotations are ordered from oldest to newest so instead of getting the first, should we get the last?

Also looping through the annotations every time we get an annotation seems inefficient. Should we create a map from workgroupId to annotation for constant time lookup?

geoffreykwan commented on src/main/java/org/wise/portal/service/peergrouping/logic/impl/PeerGroupLogicServiceImpl.java:
```diff
@@ -0,0 +1,98 @@
+  public PeerGroup createPeerGroup(Workgroup workgroup, PeerGrouping peerGrouping)
+      throws PeerGroupCreationException {
+    List<Workgroup> workgroupsInPeriod = getWorkgroupsInSamePeriod(workgroup);
+    List<Workgroup> workgroupsInPeerGroup = peerGroupDao.getWorkgroupsInPeerGroup(peerGrouping,
+        workgroup.getPeriod());
```
Maybe rename workgroupsInPeerGroup to workgroupsInAPeerGroup throughout the file because it sounds like the list is the workgroups that are in the peer group we are trying to create but it's actually the workgroups that are already in a peer group?

Likewise, maybe change workgroupsNotInPeerGroup to workgroupsNotInAPeerGroup?

hirokiterashima commented:
Thanks for the suggestions. I've made the changes. I think we can hold off on the optimization when looking up workgroup annotation until we see a performance issue, to avoid over-engineering it and making the code more complex.

geoffreykwan commented on src/main/java/org/wise/portal/service/peergrouping/logic/impl/DifferentIdeasLogicServiceImpl.java:
```diff
@@ -0,0 +1,121 @@
+    return hasWorkgroupAnnotation(workgroup, annotations)
+        && hasEnoughUnpairedMembersIdeasAnnotation(peerGrouping, annotations,
+            workgroupsNotInPeerGroup);
+  }
```
I didn't notice this file should also have the variable name workgroupsNotInPeerGroup changed to workgroupsNotInAPeerGroup. There are other files that use the variable name workgroupsNotInPeerGroup but we can change those in a different issue.

I also looked in to see how much work it would be to use a HashMap and it was pretty simple and simplified the code.
```
package org.wise.portal.service.peergrouping.logic.impl;

import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.TreeSet;

import org.apache.commons.collections4.SetUtils;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.wise.portal.dao.annotation.wise5.AnnotationDao;
import org.wise.portal.domain.group.Group;
import org.wise.portal.domain.peergrouping.PeerGrouping;
import org.wise.portal.domain.peergrouping.logic.DifferentIdeasLogic;
import org.wise.portal.domain.workgroup.Workgroup;
import org.wise.portal.service.peergroup.impl.WorkgroupWithDifferentIdeas;
import org.wise.vle.domain.annotation.wise5.Annotation;

@Service
public class DifferentIdeasLogicServiceImpl extends PeerGroupLogicServiceImpl {

  @Autowired
  private AnnotationDao<Annotation> annotationDao;

  boolean canCreatePeerGroup(Workgroup workgroup, Set<Workgroup> workgroupsNotInPeerGroup,
      PeerGrouping peerGrouping) {
    List<Annotation> annotations = getIdeaDetectedAnnotations(peerGrouping, workgroup.getPeriod());
    HashMap<Workgroup, Annotation> workgroupToAnnotation = createWorkgroupToAnnotation(annotations);
    return hasWorkgroupAnnotation(workgroup, workgroupToAnnotation)
        && hasEnoughUnpairedMembersIdeasAnnotation(peerGrouping, workgroupToAnnotation,
            workgroupsNotInPeerGroup);
  }

  private HashMap<Workgroup, Annotation> createWorkgroupToAnnotation(List<Annotation> annotations) {
    HashMap<Workgroup, Annotation> workgroupToAnnotation = new HashMap<Workgroup, Annotation>();
    for (Annotation annotation : annotations) {
      Workgroup workgroup = annotation.getToWorkgroup();
      workgroupToAnnotation.put(workgroup, annotation);
    }
    return workgroupToAnnotation;
  }

  private List<Annotation> getIdeaDetectedAnnotations(PeerGrouping peerGrouping, Group period) {
    DifferentIdeasLogic logic = new DifferentIdeasLogic(peerGrouping.getLogic());
    return annotationDao.getAnnotationsByParams(null, peerGrouping.getRun(), period, null, null,
        logic.getNodeId(), logic.getComponentId(), null, null, null, "autoScore");
  }

  private boolean hasWorkgroupAnnotation(Workgroup workgroup,
      HashMap<Workgroup, Annotation> workgroupToAnnotation) {
    return workgroupToAnnotation.containsKey(workgroup);
  }

  private Optional<Annotation> getLatestWorkgroupAnnotation(Workgroup workgroup,
      HashMap<Workgroup, Annotation> workgroupToAnnotation) {
    return Optional.of(workgroupToAnnotation.get(workgroup));
  }

  private boolean hasEnoughUnpairedMembersIdeasAnnotation(PeerGrouping peerGrouping,
      HashMap<Workgroup, Annotation> workgroupToAnnotation,
      Set<Workgroup> workgroupsNotInPeerGroup) {
    return workgroupsNotInPeerGroup.stream()
        .filter(workgroup -> hasWorkgroupAnnotation(workgroup, workgroupToAnnotation))
        .count() >= peerGrouping.getMaxMembershipCount();
  }

  Set<Workgroup> groupMembersUpToMaxMembership(Workgroup workgroup, PeerGrouping peerGrouping,
      Set<Workgroup> possibleMembers) {
    Set<Workgroup> members = new HashSet<Workgroup>();
    members.add(workgroup);
    possibleMembers.remove(workgroup);
    TreeSet<WorkgroupWithDifferentIdeas> workgroupsWithDifferentIdeas = getWorkgroupsWithDifferentIdeas(
        possibleMembers, workgroup, peerGrouping);
    Iterator<WorkgroupWithDifferentIdeas> iterator = workgroupsWithDifferentIdeas
        .descendingIterator();
    while (members.size() < peerGrouping.getMaxMembershipCount()) {
      members.add(iterator.next().getWorkgroup());
    }
    return members;
  }

  private TreeSet<WorkgroupWithDifferentIdeas> getWorkgroupsWithDifferentIdeas(
      Set<Workgroup> possibleMembers, Workgroup workgroup, PeerGrouping peerGrouping) {
    List<Annotation> annotations = getIdeaDetectedAnnotations(peerGrouping, workgroup.getPeriod());
    HashMap<Workgroup, Annotation> workgroupToAnnotation = createWorkgroupToAnnotation(annotations);
    Set<String> workgroupIdeas = getDetectedIdeas(workgroup, workgroupToAnnotation);
    return getWorkgroupsWithDifferentIdeas(possibleMembers, workgroupIdeas, workgroupToAnnotation);
  }

  private Set<String> getDetectedIdeas(Workgroup workgroup,
      HashMap<Workgroup, Annotation> workgroupToAnnotation) {
    Optional<Annotation> workgroupAnnotation = getLatestWorkgroupAnnotation(workgroup,
        workgroupToAnnotation);
    return workgroupAnnotation.isPresent() ? getDetectedIdeas(workgroupAnnotation.get())
      : new HashSet<String>();
  }

  private Set<String> getDetectedIdeas(Annotation annotation) {
    try {
      JSONObject dataJson = new JSONObject(annotation.getData());
      JSONArray ideas = dataJson.getJSONArray("ideas");
      Set<String> detectedIdeas = new HashSet<String>();
      for (int i = 0; i < ideas.length(); i++) {
        JSONObject idea = ideas.getJSONObject(i);
        if (idea.getBoolean("detected")) {
          detectedIdeas.add(idea.getString("name"));
        }
      }
      return detectedIdeas;
    } catch (JSONException e) {
      return new HashSet<String>();
    }
  }

  private TreeSet<WorkgroupWithDifferentIdeas> getWorkgroupsWithDifferentIdeas(
      Set<Workgroup> possibleMembers, Set<String> workgroupIdeas,
      HashMap<Workgroup, Annotation> workgroupToAnnotation) {
    TreeSet<WorkgroupWithDifferentIdeas> workgroups = new TreeSet<WorkgroupWithDifferentIdeas>();
    for (Workgroup possibleMember : possibleMembers) {
      if (hasWorkgroupAnnotation(possibleMember, workgroupToAnnotation)) {
        Set<String> possibleMemberIdeas = getDetectedIdeas(possibleMember, workgroupToAnnotation);
        Set<String> differentIdeas = SetUtils.disjunction(workgroupIdeas, possibleMemberIdeas);
        workgroups.add(new WorkgroupWithDifferentIdeas(possibleMember, differentIdeas.size()));
      }
    }
    return workgroups;
  }
}

```

geoffreykwan commented on src/main/java/org/wise/portal/service/peergrouping/logic/impl/DifferentIdeasLogicServiceImpl.java:
```diff
@@ -0,0 +1,121 @@
+    } catch (JSONException e) {
+      return new HashSet<String>();
+    }
+  }
+
+  private TreeSet<WorkgroupWithDifferentIdeas> getWorkgroupsWithDifferentIdeas(
+      Set<Workgroup> possibleMembers, Set<String> workgroupIdeas, List<Annotation> annotations) {
+    TreeSet<WorkgroupWithDifferentIdeas> workgroups = new TreeSet<WorkgroupWithDifferentIdeas>();
+    for (Workgroup possibleMember : possibleMembers) {
+      if (hasWorkgroupAnnotation(possibleMember, annotations)) {
+        Set<String> possibleMemberIdeas = getDetectedIdeas(possibleMember, annotations);
+        Set<String> differentIdeas = SetUtils.disjunction(workgroupIdeas, possibleMemberIdeas);
+        workgroups.add(new WorkgroupWithDifferentIdeas(possibleMember, differentIdeas.size()));
+      }
+    }
+    return workgroups;
+  }
+}
```
I noticed that the TreeSet can't contain duplicates which means for a given number of different ideas, it will only have one workgroup.

For example
Workgroup aa has idea 1 and is the one making the request to pair up
Workgroup bb has idea 2
Workgroup cc has ideas 3, 4
Workgroup dd has ideas 3, 4

The workgroups TreeSet will end up with only bb and cc. It will not contain dd because cc and dd have the same number of differentIdeas compared to aa.

hirokiterashima commented:
Good catch about workgroups with same number of differentDetectedIdeas. I've fixed it by modifying WorkgroupWithDifferentIdeas.compareTo().

Also cleaned up the code using HashSet and renamed workgroupsNotInPeerGroup => workgroupsNotInAPeerGroup

geoffreykwan reviewed (approved):
Looks good.

geoffreykwan commented:
:tada: This PR is included in version 1.0.0 :tada:

The release is available on [GitHub release](https://github.com/WISE-Community/WISE-API/releases/tag/v1.0.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
